### PR TITLE
Update functions to middlewares and final format is also http handler

### PR DESCRIPTION
**Pull request by LaurenceJJones**
Allow option for #12 
Allow customisation of microtik listname @williamdes

Convert other functions into middlewares so early return is simpler

Memory consumption is a little concern due to filters, however, stress testing only shown a 6mb increase. 

When I mean stress test that was 1k request per second over 10 minutes.

### CODE REVIEW

**Review by williamdes — APPROVED** [review]
Awesome !

**Comment by williamdes** [comment]
Also update the `docker/README.md` ?

File: formatters.go [hunk 1]
```diff
@@ -2,43 +2,51 @@ package main
...
 			*decision.Scenario,
 			*decision.Duration,
 		)
 	}
 	sort.Strings(ips)
-	return "/ip firewall address-list remove [find list=CrowdSec]\n" +
-		"/ipv6 firewall address-list remove [find list=CrowdSec]\n" +
-		strings.Join(ips, "\n")
+	w.Write([]byte(fmt.Sprintf("/ip firewall address-list remove [find list=%s]\n", listName) +
+		fmt.Sprintf("/ipv6 firewall address-list remove [find list=%s]\n", listName) +
```
Comment (williamdes) [inline formatters.go:50]: You need to remove this ipv6 line if it's v4 only mode because it may indicate that the user does not have v6 module enabled

Comment (LaurenceJJones) [reply formatters.go]: I have made it to remove v4 if onlyipv6 and remove v6 if only ipv4?

Comment (williamdes) [reply formatters.go]: Well, yes. But technically ipv4 is always enabled. But it would be cleaner

**Review by williamdes — APPROVED** [review]
🚀

File: docker/README.md [hunk 2]
```diff
@@ -185,3 +201,15 @@ Example:
 /ip firewall address-list add list=CrowdSec address=1.2.3.4 comment="crowdsecurity/ssh-bf for 152h40m24.308868973s"
 /ip firewall address-list add list=CrowdSec address=4.3.2.1 comment="crowdsecurity/postfix-spam for 166h40m25.280338424s"/ipv6 firewall address-list add list=CrowdSec address=2001:470:1:c84::17 comment="crowdsecurity/ssh-bf for 165h13m42.405449876s"
 ```
+
+#### microtik query parameters
+
+`?listname=foo` - Set the list name to `foo`
```
Comment (williamdes) [inline docker/README.md:207]:
```suggestion
`?listname=foo` - Set the list name to `foo`, by default `listname` is set to `CrowdSec`
```

**Review by williamdes — APPROVED** [review]

**Review by williamdes — APPROVED** [review]
Maybe merge #24 and rebase onto so we try the new workflows on this one ?

**Comment by LaurenceJJones** [comment]
> Maybe merge #24 and rebase onto so we try the new workflows on this one ?

I ask another team member for a review, since they better at CI and I might miss something since I been staring at code for last 2 hours 😆 

Appreciate your input on this PR, I will push for release after this PR been merged as its a major change.

**Comment by williamdes** [comment]
If you have some mac of guys ask them if the Linux binaries we distribute run on their workstation or if I should add Darwin binaries too

**Comment by LaurenceJJones** [comment]
Just looking at optimisations why do we need to create an ips within the microtik formatter? plus do we care about sorting strings? seems abit pointless? unless there a reason?

**Review by williamdes — APPROVED** [review]
That's okay to have them out of order, maybe it's some kind of router optimisation someone had did. I just copied it as it made pretty output that kind of deterministic
But that's okay to have it unsorted, or maybe add another query option ?
Maybe some people/software group the ips ?

PS: last commit is not GPG signed 🤔

**Comment by LaurenceJJones** [comment]
> PS: last commit is not GPG signed thinking

I messed up the rebase 😮‍💨 

It doesn't matter now, I now made it so when we fetch the decisions it sorts them so it should be same order everytime.

**Comment by williamdes** [comment]
> > PS: last commit is not GPG signed thinking
> 
> I messed up the rebase 😮‍💨
> 
> It doesn't matter now, I now made it so when we fetch the decisions it sorts them so it should be same order everytime.

Awesome!
Good for the then

**Comment by LaurenceJJones** [comment]
```
This is ApacheBench, Version 2.3 <$Revision: 1901567 $>
Copyright 1996 Adam Twiss, Zeus Technology Ltd, http://www.zeustech.net/
Licensed to The Apache Software Foundation, http://www.apache.org/

Benchmarking localhost (be patient)
Completed 1000 requests
Completed 2000 requests
Completed 3000 requests
Completed 4000 requests
Completed 5000 requests
Completed 6000 requests
Completed 7000 requests
Completed 8000 requests
Completed 9000 requests
Completed 10000 requests
Finished 10000 requests


Server Software:
Server Hostname:        localhost
Server Port:            4141

Document Path:          /
Document Length:        3704276 bytes

Concurrency Level:      10
Time taken for tests:   45.741 seconds
Complete requests:      10000
Failed requests:        0
Keep-Alive requests:    0
Total transferred:      37043730000 bytes
HTML transferred:       37042760000 bytes
Requests per second:    218.62 [#/sec] (mean)
Time per request:       45.741 [ms] (mean)
Time per request:       4.574 [ms] (mean, across all concurrent requests)
Transfer rate:          790874.24 [Kbytes/sec] received

Connection Times (ms)
min  mean[+/-sd] median   max
Connect:        0    0   0.0      0       0
Processing:    26   46   6.5     45      73
Waiting:       14   24   4.7     24      40
Total:         26   46   6.5     45      73

Percentage of the requests served within a certain time (ms)
50%     45
66%     48
75%     50
80%     51
90%     54
95%     58
98%     60
99%     62
100%     73 (longest request)
```

**Comment by williamdes** [comment]
Is sorting responsible for the overhead?

**Comment by LaurenceJJones** [comment]
> Is sorting responsible for the overhead?

Microtik no sort 1.7ms faster per request
```
This is ApacheBench, Version 2.3 <$Revision: 1901567 $>
Copyright 1996 Adam Twiss, Zeus Technology Ltd, http://www.zeustech.net/
Licensed to The Apache Software Foundation, http://www.apache.org/

Benchmarking localhost (be patient)
Completed 1000 requests
Completed 2000 requests
Completed 3000 requests
Completed 4000 requests
Completed 5000 requests
Completed 6000 requests
Completed 7000 requests
Completed 8000 requests
Completed 9000 requests
Completed 10000 requests
Finished 10000 requests


Server Software:
Server Hostname:        localhost
Server Port:            4141

Document Path:          /
Document Length:        3578225 bytes

Concurrency Level:      10
Time taken for tests:   28.034 seconds
Complete requests:      10000
Failed requests:        0
Keep-Alive requests:    0
Total transferred:      35783220000 bytes
HTML transferred:       35782250000 bytes
Requests per second:    356.71 [#/sec] (mean)
Time per request:       28.034 [ms] (mean)
Time per request:       2.803 [ms] (mean, across all concurrent requests)
Transfer rate:          1246511.39 [Kbytes/sec] received

Connection Times (ms)
min  mean[+/-sd] median   max
Connect:        0    0   0.0      0       0
Processing:    18   28   3.9     28      45
Waiting:        2    4   1.4      4       9
Total:         18   28   3.9     28      45

Percentage of the requests served within a certain time (ms)
50%     28
66%     30
75%     31
80%     31
90%     33
95%     35
98%     36
99%     37
100%     45 (longest request)
```

plain_text no sort 1.8ms faster
```
This is ApacheBench, Version 2.3 <$Revision: 1901567 $>
Copyright 1996 Adam Twiss, Zeus Technology Ltd, http://www.zeustech.net/
Licensed to The Apache Software Foundation, http://www.apache.org/

Benchmarking localhost (be patient)
Completed 1000 requests
Completed 2000 requests
Completed 3000 requests
Completed 4000 requests
Completed 5000 requests
Completed 6000 requests
Completed 7000 requests
Completed 8000 requests
Completed 9000 requests
Completed 10000 requests
Finished 10000 requests


Server Software:
Server Hostname:        localhost
Server Port:            4141

Document Path:          /
Document Length:        410401 bytes

Concurrency Level:      10
Time taken for tests:   12.260 seconds
Complete requests:      10000
Failed requests:        0
Keep-Alive requests:    0
Total transferred:      4104980000 bytes
HTML transferred:       4104010000 bytes
Requests per second:    815.64 [#/sec] (mean)
Time per request:       12.260 [ms] (mean)
Time per request:       1.226 [ms] (mean, across all concurrent requests)
Transfer rate:          326972.88 [Kbytes/sec] received

Connection Times (ms)
min  mean[+/-sd] median   max
Connect:        0    0   0.0      0       0
Processing:     6   12   2.6     12      31
Waiting:        2    4   1.4      4      13
Total:          6   12   2.6     12      31

Percentage of the requests served within a certain time (ms)
50%     12
66%     13
75%     14
80%     15
90%     16
95%     16
98%     17
99%     18
100%     31 (longest request)
```

obviously sorting for every request will cause some overhead unless we provide some sort of caching but at that point it might be better to put a cache infront of the blocklist bouncer.

**Comment by williamdes** [comment]
At this point I'd argue it's a good idea to add a way to disable sorting just like IPv6
This way people can have a grain of salt more performance
But it's not that much gain

**Comment by LaurenceJJones** [comment]
> At this point I'd argue it's a good idea to add a way to disable sorting just like IPv6 This way people can have a grain of salt more performance But it's not that much gain

Added option for `?nosort` the default will be to sort and if they want to heavily optimise then they can find this parameter

**Review by blotus — APPROVED** [review]

**Pull request merged**